linkerd/buffer/src/dispatch.rs

```diff
@@ -0,0 +1,136 @@
+        loop {
+            let ready = match self.inner.as_mut() {
+                Some(inner) => inner.poll_ready(),
+                None => {
+                    // This is safe because ready.poll_close has returned NotReady.
+                    return Ok(Async::NotReady);
+                }
+            };
+
+            match ready {
+                // If it's not ready, wait for it..
+                Ok(Async::NotReady) => {
+                    return_ready_if!(self.ready.broadcast(Ok(Async::NotReady)).is_err());
+
+                    trace!("Waiting for inner service");
+                    return Ok(Async::NotReady);
+                }
+
+                // If the service fails, propagate the failure to all pending
+                // requests and then complete.
+                Err(error) => {
+                    let shared = ServiceError(Arc::new(error.into()));
+                    trace!(%shared, "Inner service failed");
+
+                    // First, notify services of the readiness change to prevent new requests from
+                    // being buffered.
+                    let is_active = self.ready.broadcast(Err(shared.clone())).is_ok();
+
+                    // Propagate the error to all in-
+                    while let Ok(Async::Ready(Some(InFlight { tx, .. }))) = self.rx.poll() {
+                        let _ = tx.send(Err(shared.clone().into()));
+                    }
+
+                    // Drop the inner Service to free its resources. It won't be used again.
+                    self.inner = None;
+
+                    // Ensure the task remains active until all services have observed the error.
+                    return_ready_if!(!is_active);
+
+                    // This is safe because ready.poll_close has returned NotReady.
+                    return Ok(Async::NotReady);
+                }
+
+                // If inner service can receive requests, start polling the channel.
+                Ok(Async::Ready(())) => {
+                    return_ready_if!(self.ready.broadcast(Ok(Async::Ready(()))).is_err());
+                    trace!("Ready for requests");
+                }
+            }
+
+            // The inner service is ready, so poll for new requests.
+            match self.rx.poll() {
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+
+                // All senders have been dropped, complete.
+                Err(_) | Ok(Async::Ready(None)) => return_ready!(),
+
+                // If a request was ready return it to the caller.
+                Ok(Async::Ready(Some(InFlight { request, tx }))) => {
+                    trace!("Dispatching a request");
+                    let fut = self
+                        .inner
+                        .as_mut()
+                        .expect("Service must not be dropped")
+                        .call(request);
+                    let _ = tx.send(Ok(fut));
```

Comment:
how often do we think requesters will hang up after requests are dispatched? is this something that we would want to record a TRACE/DEBUG event for?

Reply:
This will happen whenever a request is canceled (i.e. the client resets the stream or drops the connection). Could go either way